File: practice/longest_substring_with_k_distinct.js
function longest_substring_with_k_distinct(str, s) {
  const wordFrequency = {};
  let maxLen = 0;
  let windowstart = 0;
  for (let i = 0; i < str.length; i++) {
    const rightchar = str[i];
    if (!(rightchar in wordFrequency)) {
      wordFrequency[rightchar] = 0;
    }
    wordFrequency[rightchar]++;
    while (Object.keys(wordFrequency).length > s) {
      const leftchar = str[windowstart];
      wordFrequency[leftchar]--;
      if (wordFrequency[leftchar] === 0) {
        delete wordFrequency[leftchar];
      }
      windowstart++;
    }
    maxLen = Math.max(maxLen, i - windowstart + 1);
  }
  return maxLen;
}

console.log(longest_substring_with_k_distinct("araaci", 2));
console.log(longest_substring_with_k_distinct("araaci", 1));

File: educative/go/longest_substring_with_k_distinct.go
package main
import "fmt"

// find the longest substring with only k different characters
func longest_substring_with_k_distinct(str string, k int) int {
	var print = fmt.Println
	print("starting algorithm")
	var maxlen = 0
	var start = 0;
	var frequencyChars = make(map[string]int);
	for end, c := range str{
		var rightchar = string(c);
		frequencyChars[rightchar] += 1;
		for len(frequencyChars) > k{
			var leftchar = string(str[start]);
			frequencyChars[leftchar] -= 1;
			if frequencyChars[leftchar] == 0{
				delete(frequencyChars, leftchar);
			}
			start += 1;
		}
		maxlen = max(maxlen, end - start + 1);
	}
	return maxlen
}

func max(n, m int) int {
	if n > m {
			return n
	}
	return m
}


func main(){
	fmt.Println(longest_substring_with_k_distinct("araaci", 2))
	fmt.Println(longest_substring_with_k_distinct("araaci", 1))
	fmt.Println(longest_substring_with_k_distinct("cbbebi", 3))
}